lmnet/lmnet/blocks.py

```diff
@@ -126,3 +128,192 @@ def lmnet_block(
+    else:
+        raise ValueError("data format must be 'NCHW' or 'NHWC'. got {}.".format(data_format))
+
+    with tf.variable_scope(name):
+        conved = tf.layers.conv2d(
+            inputs,
+            filters=filters,
+            kernel_size=kernel_size,
+            padding='SAME',
+            use_bias=False,
+            kernel_initializer=tf.contrib.layers.variance_scaling_initializer(),  # he initializer
+            data_format=channel_data_format,
+            kernel_regularizer=tf.contrib.layers.l2_regularizer(weight_decay_rate),
+        )
+
+        batch_normed = tf.contrib.layers.batch_norm(
+            conved,
+            decay=batch_norm_decay,
+            updates_collections=None,
+            is_training=is_training,
+            center=True,
+            scale=True,
+            data_format=data_format,
+            )
+
+        if activation:
+            output = activation(batch_normed)
+        else:
+            output = batch_normed
+
+        if enable_detail_summary:
+            tf.summary.histogram('conv_output', conved)
+            tf.summary.histogram('batch_norm_output', batch_normed)
+            tf.summary.histogram('output', output)
+
+        return output
+
+
+def _dense_layer_conv_bn_act(
+        name,
+        inputs,
+        growth_rate,
+        bottleneck_rate,
+        weight_decay_rate,
+        is_training,
+        activation,
+        batch_norm_decay,
+        data_format,
+        enable_detail_summary,
+):
+    """Densenet lyaer
+
+    In order to fast execute for quantization, use order of layer
+    convolution -> batch norm -> activation instead of paper original's batch norm -> activation -> convolution.
+    """
+    bottleneck_channel = growth_rate * bottleneck_rate
+
+    with tf.variable_scope(name):
+
+        output_1x1 = conv_bn_act(
+            "bottleneck_1x1",
+            inputs,
+            filters=bottleneck_channel,
+            kernel_size=1,
+            weight_decay_rate=weight_decay_rate,
+            is_training=is_training,
+            activation=activation,
+            batch_norm_decay=batch_norm_decay,
+            data_format=data_format,
+            enable_detail_summary=enable_detail_summary,
+        )
+
+        output_3x3 = conv_bn_act(
+            "conv_3x3",
+            output_1x1,
+            filters=growth_rate,
+            kernel_size=3,
+            weight_decay_rate=weight_decay_rate,
+            is_training=is_training,
+            activation=activation,
+            batch_norm_decay=batch_norm_decay,
+            data_format=data_format,
+            enable_detail_summary=enable_detail_summary,
+        )
+
+        if data_format == "NHWC":
+            concat_axis = -1
+        if data_format == "NCHW":
+            concat_axis = 1
+
+        output = tf.concat([inputs, output_3x3], axis=concat_axis)
+
+        if enable_detail_summary:
+            tf.summary.histogram('output', output)
+
+    return output
+
+
+def densenet(
```

Comment:
`densnet` looks like name of the network.
In the original paper, authers called `dense block` ?

Reply:
Yes, the `densenet()` is `Dense block` in original paper.

Which one is best ?
1. `densenet()`,  `mobilenetv2()`
2. `dense_block()`, `mobilev2_block()`
3. `dense()`, `mobilev2()` 

No. 2 is not good for me, the function is obviously block, because this function is under `blocks.py`, It is duplicate.
No. 3 `dense` is confuse with `fully connect` (some library name `dense` as `fully connect`)  .
That's why I choose No.1

But just now, I think the name `densenet_group()` is better than `densenet()`. Because
most famous `Residual block` abstraction level is same as current `_dense_layer_conv_bn_act` . `Residual block` and  `_dense_layer_conv_bn_act()` are composed from some convolution and batch norm, etc. `densenet()` are composed from some the `_dense_layer_conv_bn_act()` , It's abstraction level is difference from `Residual block`.

Reply:
> No. 2 is not good for me, the function is obviously block, because this function is under blocks.py, It is duplicate.

OK. There is `lmnet_block` in this script. Thats why I commented.

`densenet_group` looks good name.

Reply:
Fixed in https://github.com/blue-oil/blueoil/pull/212/commits/1260d4f50a1d3850cedfdc6929cabce496f3f19e